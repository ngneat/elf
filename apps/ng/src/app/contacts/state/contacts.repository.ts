import { Injectable } from '@angular/core';
import { createStore } from '@ngneat/elf';
import { addEntities, withEntities } from '@ngneat/elf-entities';
import {
  PaginationData,
  selectCurrentPage,
  selectCurrentPageEntities,
  selectPaginationData,
  setCurrentPage,
  setPage,
  updatePaginationData,
  withPagination,
} from '@ngneat/elf-pagination';

interface Contact {
  id: number;
  email: string;
  name: string;
  address: string;
}

const store = createStore(
  { name: 'contacts' },
  withPagination(),
  withEntities<Contact>()
);

@Injectable({ providedIn: 'root' })
export class ContactsRepository {
  activePage$ = store.pipe(selectCurrentPage());
  paginationData$ = store.pipe(selectPaginationData());
  activePageContacts$ = store.pipe(selectCurrentPageEntities());

  setActivePage(id: Contact['id']) {
    store.update(setCurrentPage(id));
  }

  addContacts(page: number, contacts: PaginationData & { data: Contact[] }) {
    const { data, ...pagination } = contacts;

    store.update(
      addEntities(data),
      updatePaginationData(pagination),
      setPage(
        page,
        data.map((c) => c.id)
      )
    );
  }

  get store() {
    return store;
  }
}
