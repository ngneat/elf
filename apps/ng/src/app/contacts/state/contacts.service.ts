import { Injectable } from '@angular/core';
import { getData } from '../contacts-data';
import { timer } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { ContactsRepository } from './contacts.repository';
import { skipWhilePageExists } from '@ngneat/elf-pagination';

@Injectable({ providedIn: 'root' })
export class ContactsService {
  constructor(private repo: ContactsRepository) {}

  getContacts(page: number) {
    return timer(300).pipe(
      map(() => getData({ page })),
      tap((res) => this.repo.addContacts(page, res)),
      skipWhilePageExists(this.repo.store, page)
    );
  }
}
