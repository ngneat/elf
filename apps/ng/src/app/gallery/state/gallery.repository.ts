import { Injectable } from '@angular/core';
import { createStore } from '@ngneat/elf';
import {
  addEntities,
  selectAllEntities,
  selectEntities,
  UIEntitiesRef,
  unionEntities,
  updateEntities,
  withEntities,
  withUIEntities,
} from '@ngneat/elf-entities';

export interface GalleryItem {
  id: number;
  title: string;
  description: string;
  image: string;
}

export interface GalleryUIItem {
  id: number;
  open: boolean;
}

const store = createStore(
  { name: 'gallery' },
  withEntities<GalleryItem>(),
  withUIEntities<GalleryUIItem>()
);

@Injectable({ providedIn: 'root' })
export class GalleryRepository {
  items$ = this.store
    .combine({
      entities: store.pipe(selectAllEntities()),
      UIEntities: store.pipe(selectEntities({ ref: UIEntitiesRef })),
    })
    .pipe(unionEntities());

  addGalleryItems(galleryItems: GalleryItem[]) {
    const uiGalleryItems = galleryItems.map(({ id }) => ({ id, open: false }));
    store.update(
      addEntities(galleryItems),
      addEntities(uiGalleryItems, { ref: UIEntitiesRef })
    );
  }

  get store() {
    return store;
  }

  toggleItemOpen(galleryItemId: number) {
    this.store.update(
      updateEntities(galleryItemId, (item) => ({ ...item, open: !item.open }), {
        ref: UIEntitiesRef,
      })
    );
  }
}
