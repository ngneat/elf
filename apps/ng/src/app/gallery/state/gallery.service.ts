import { Injectable } from '@angular/core';
import { getData } from '../gallery-data';
import { timer } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { GalleryRepository } from './gallery.repository';

@Injectable({ providedIn: 'root' })
export class GalleryService {
  constructor(private repo: GalleryRepository) {}

  getGalleryItems() {
    return timer(300).pipe(
      map(() => getData()),
      tap((res) => this.repo.addGalleryItems(res))
    );
  }
}
