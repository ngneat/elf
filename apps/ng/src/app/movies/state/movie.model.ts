export type Movie = {
  id: number | string;
  title: string;
  genres: (string | number)[];
  actors: (string | number)[];
};

export interface FullMovie {
  id: number | string;
  title: string;
  genres: { id: number | string; name: string }[];
  actors: { id: number | string; name: string }[];
}
