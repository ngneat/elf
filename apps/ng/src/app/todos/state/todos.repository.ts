import { Injectable } from '@angular/core';
import { createStore, select, withProps } from '@ngneat/elf';
import {
  addEntities,
  selectAllEntities,
  selectAllEntitiesApply,
  selectEntityByPredicate,
  selectManyByPredicate,
  updateEntities,
  withEntities,
} from '@ngneat/elf-entities';
import { switchMap } from 'rxjs/operators';
import { write } from '../../store/mutations';

interface Todo {
  id: number;
  title: string;
  completed: boolean;
}

export interface TodosProps {
  filter: 'ALL' | 'ACTIVE' | 'COMPLETED';
}

const store = createStore(
  { name: 'todos' },
  withEntities<Todo>(),
  withProps<TodosProps>({ filter: 'ALL' })
);

@Injectable({ providedIn: 'root' })
export class TodosRepository {
  todos$ = store.pipe(selectAllEntities());
  filter$ = store.pipe(select((state) => state.filter));

  visibleTodos$ = this.filter$.pipe(
    switchMap((filter) => {
      return store.pipe(
        selectAllEntitiesApply({
          filterEntity({ completed }) {
            if (filter === 'ALL') return true;
            return filter === 'COMPLETED' ? completed : !completed;
          },
        })
      );
    })
  );

  selectByCompletedState(completed: Todo['completed']) {
    return store.pipe(
      selectManyByPredicate((entity) => entity.completed === completed)
    );
  }

  selectFirstCompletedTitle() {
    return store.pipe(
      selectEntityByPredicate((entity) => entity.completed, { pluck: 'title' })
    );
  }

  addTodo(title: Todo['title']) {
    store.update(addEntities({ id: 1, title, completed: false }));
    store.update(
      addEntities([
        { id: 1, title, completed: false },
        { id: 2, title, completed: false },
      ])
    );
  }

  updateFilter(filter: TodosProps['filter']) {
    store.update(
      write((state) => {
        state.filter = filter;
      })
    );
  }

  updateCompleted(id: Todo['id']) {
    store.update(
      updateEntities(id, (entity) => ({
        ...entity,
        completed: !entity.completed,
      }))
    );
  }
}
