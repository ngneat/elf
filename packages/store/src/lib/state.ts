type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never;
type Merge<State extends any[], Key extends PropertyKey> = UnionToIntersection<
  State[number][Key]
>;

export type PropsFactory<Props, Config> = { props: Props; config: Config };
export type EmptyConfig = undefined;

export function createState<S extends PropsFactory<any, any>[]>(
  ...propsFactories: S
): { state: Merge<S, 'props'>; config: Merge<S, 'config'> } {
  const result = { config: {}, state: {} };

  for (const { config, props } of propsFactories) {
    Object.assign(result.config, config);
    Object.assign(result.state, props);
  }

  return result as { state: Merge<S, 'props'>; config: Merge<S, 'config'> };
}
