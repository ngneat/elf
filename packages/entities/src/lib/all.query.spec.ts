import {
  createEntitiesStore,
  createTodo,
  createUIEntityStore,
  createUITodo,
  Todo,
} from '@ngneat/elf-mocks';
import { addEntities } from './add.mutation';
import {
  selectAllEntities,
  selectAllEntitiesApply,
  selectEntities,
} from './all.query';
import { UIEntitiesRef } from './entity.state';
import { expectTypeOf } from 'expect-type';
import { updateEntities } from '@ngneat/elf-entities';

describe('selectAllEntities', () => {
  let store: ReturnType<typeof createEntitiesStore>;

  beforeEach(() => {
    store = createEntitiesStore();
  });

  it('should select all', () => {
    let count = 1;

    store.pipe(selectAllEntities()).subscribe((value) => {
      expect(value).toMatchSnapshot(`calls: ${count++}`);
    });

    store.update(addEntities(createTodo(2)));
  });

  it('should work with ref', () => {
    let count = 1;

    const store = createUIEntityStore();

    store.pipe(selectAllEntities({ ref: UIEntitiesRef })).subscribe((value) => {
      expect(value).toMatchSnapshot(`calls: ${count++}`);
    });

    store.update(
      addEntities([createUITodo(1), createUITodo(2)], { ref: UIEntitiesRef })
    );
  });

  it('should observe entities object', () => {
    store.pipe(selectEntities()).subscribe((v) => {
      expect(v).toMatchSnapshot('should be an object');
    });

    store.update(addEntities(createTodo(1)));
  });

  describe('selectAllEntitiesApply', () => {
    const store = createEntitiesStore();

    store.update(
      addEntities(createTodo(1)),
      addEntities(createTodo(2)),
      updateEntities(1, { completed: true })
    );

    it('should map', () => {
      store
        .pipe(selectAllEntitiesApply({ mapEntity: (e) => e.title }))
        .subscribe((v) => {
          expectTypeOf(v).toEqualTypeOf<string[]>();
          expect(v).toStrictEqual(['todo 1', 'todo 2']);
        });
    });

    it('should filter', () => {
      store
        .pipe(selectAllEntitiesApply({ filterEntity: (e) => e.completed }))
        .subscribe((v) => {
          expectTypeOf(v).toEqualTypeOf<Todo[]>();
          expect(v).toStrictEqual([
            { id: 1, title: 'todo 1', completed: true },
          ]);
        });
    });

    it('should filter and then map', () => {
      store
        .pipe(
          selectAllEntitiesApply({
            mapEntity: (e) => e.title,
            filterEntity: (e) => e.completed,
          })
        )
        .subscribe((v) => {
          expectTypeOf(v).toEqualTypeOf<string[]>();
          expect(v).toStrictEqual(['todo 1']);
        });
    });

    it('should filter and then map entities ref', () => {
      const store = createUIEntityStore();

      store.update(
        addEntities(createUITodo(1), { ref: UIEntitiesRef }),
        addEntities(createUITodo(2), { ref: UIEntitiesRef }),
        updateEntities(1, { open: true }, { ref: UIEntitiesRef })
      );

      store
        .pipe(
          selectAllEntitiesApply({
            mapEntity: (e) => e.id,
            filterEntity: (e) => e.open,
            ref: UIEntitiesRef,
          })
        )
        .subscribe((v) => {
          expectTypeOf(v).toEqualTypeOf<number[]>();
          expect(v).toStrictEqual([1]);
        });
    });
  });
});
