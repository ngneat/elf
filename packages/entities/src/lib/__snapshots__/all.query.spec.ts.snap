// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectAllEntities should observe entities object: should be an object 1`] = `Object {}`;

exports[`selectAllEntities should observe entities object: should be an object 2`] = `
Object {
  "1": Object {
    "completed": false,
    "id": 1,
    "title": "todo 1",
  },
}
`;

exports[`selectAllEntities should select all: calls: 1 1`] = `Array []`;

exports[`selectAllEntities should select all: calls: 2 1`] = `
Array [
  Object {
    "completed": false,
    "id": 2,
    "title": "todo 2",
  },
]
`;

exports[`selectAllEntities should work with ref: calls: 1 1`] = `Array []`;

exports[`selectAllEntities should work with ref: calls: 2 1`] = `
Array [
  Object {
    "id": 1,
    "open": false,
  },
  Object {
    "id": 2,
    "open": false,
  },
]
`;
